hail/python/hail/methods/statgen.py

```diff
@@ -444,43 +437,42 @@ def select_array_indices(hl_array, indices):
 
     ht = ht_local.transmute(**{entries_field_name: ht_local[entries_field_name][x_field_name]})
 
-    def make_one_cov_matrix(ys_and_covs_to_keep):
-        return hl.nd.array(ys_and_covs_to_keep.map(lambda struct: array_from_struct(struct, cov_field_names))) \
-            if cov_field_names else hl.nd.zeros((hl.len(ys_and_covs_to_keep), 0))
-
-    def make_one_y_matrix(ys_and_covs_to_keep, one_y_field_name_set):
-        return hl.nd.array(ys_and_covs_to_keep.map(lambda struct: array_from_struct(struct, one_y_field_name_set)))
-
     def setup_globals(ht):
-        all_covs_defined = ht[sample_field_name].map(lambda sample_struct: all_defined(sample_struct, cov_field_names))
-        list_of_ys_and_covs_to_keep_with_indices = [
-            hl.enumerate(ht[sample_field_name]).filter(
-                lambda struct_with_index:
-                all_covs_defined[struct_with_index[0]] & all_defined(struct_with_index[1], one_y_field_name_set))
-            for one_y_field_name_set in y_field_names
-        ]
+        # cov_arrays is per sample, then per cov.
+        if covariates:
+            ht = ht.annotate_globals(cov_arrays=ht[sample_field_name].map(lambda sample_struct: [sample_struct[cov_name] for cov_name in cov_field_names]))
+        else:
+            ht = ht.annotate_globals(cov_arrays=ht[sample_field_name].map(lambda sample_struct: hl.empty_array(hl.tfloat64)))
 
-        list_of_ys_and_covs_to_keep = [inner_list.map(lambda pair: pair[1]) for inner_list in list_of_ys_and_covs_to_keep_with_indices]
-        kept_samples = [inner_list.map(lambda pair: pair[0]) for inner_list in list_of_ys_and_covs_to_keep_with_indices]
+        ht = ht.annotate_globals(
+            y_arrays_per_group=[ht[sample_field_name].map(lambda sample_struct: [sample_struct[y_name] for y_name in one_y_field_name_set]) for one_y_field_name_set in y_field_name_groups]
+        )
+        all_covs_defined = ht.cov_arrays.map(lambda sample_covs: sample_covs.all(lambda a: hl.is_defined(a)))
 
-        cov_nds = hl.array([make_one_cov_matrix(ys_and_covs_to_keep) for ys_and_covs_to_keep in list_of_ys_and_covs_to_keep])
+        def make_idx_ys_covs_struct(sample_ys):
+            # sample_ys is an array of samples, with each element being an array of the y_values
+            return hl.enumerate(sample_ys).filter(
+                lambda idx_and_y_values: all_covs_defined[idx_and_y_values[0]] & no_missing(idx_and_y_values[1])
+            ).map(lambda idx_and_y_values: hl.struct(idx=idx_and_y_values[0], ys=idx_and_y_values[1], covs=ht.cov_arrays[idx_and_y_values[0]]))
 
-        y_nds = hl.array([make_one_y_matrix(ys_and_covs_to_keep, one_y_field_name_set)
-                          for ys_and_covs_to_keep, one_y_field_name_set in zip(list_of_ys_and_covs_to_keep, y_field_names)])
+        kept_data_per_y_group = ht.y_arrays_per_group.map(make_idx_ys_covs_struct)
+        kept_samples = kept_data_per_y_group.idx
```

Comment:
I was trying to write things in terms of maps with the thinking that it might be possible for these things to stay as streams, rather than manifesting multiple arrays in memory. I figured our compiler would do smarter things with a map than with looping over the indices. I'll try this though.

Reply:
You're basically trying to do a single map over `y_arrays_per_group` to simultaneously define three arrays (`kept_samples`, `kept_data_per_y_group.ys`, and `kept_data_per_y_group.covs`). But we don't have any way to do that with streams, so it will have to materialize an intermediate array no matter what, or just iterate over `y_arrays_per_group` three times. I think `kept_samples` is probably the best intermediate thing to materialize. And the things you're indexing into are already materialized arrays, so no loss there.